sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -340,7 +340,8 @@ class Dataset[T] private[sql](
    */
   // This is declared with parentheses to prevent the Scala compiler from treating
   // `ds.toDF("1")` as invoking this toDF and then apply on the returned DataFrame.
-  def toDF(): DataFrame = new Dataset[Row](sparkSession, queryExecution, RowEncoder(schema))
+  def toDF(): DataFrame =
+    new Dataset[Row](sparkSession, queryExecution, RowEncoder(logicalPlan.schema))
```

Comment:
Another option here is to add a projection to make sure column order of the result DataFrame always conform to schema of the Dataset.